### PR TITLE
[WIP] io: update Reader interface

### PR DESCRIPTION
**Pull request by kt3k**
This PR updates the Reader interface based on the change introduced by https://github.com/denoland/deno/pull/2591

### CODE REVIEW

File: io/bufio.ts [hunk 1]
```diff
@@ -84,14 +80,15 @@ export class BufReader implements Reader {
 
     // Read new data: try a limited number of times.
     for (let i = MAX_CONSECUTIVE_EMPTY_READS; i > 0; i--) {
-      let rr: ReadResult = await this.rd.read(this.buf.subarray(this.w));
-      assert(rr.nread >= 0, "negative read");
-      this.w += rr.nread;
-      if (rr.eof) {
+      const rr = await this.rd.read(this.buf.subarray(this.w));
+      const nread = rr === Deno.EOF ? 0 : rr;
```
Comment (piscisaureus) [inline io/bufio.ts]: You wouldn't need this if you moved the `assert(nread >= 0)` *after* the `if (rr ===  EOF) { return}`.

File: io/bufio.ts [hunk 2]
```diff
@@ -144,17 +142,18 @@ export class BufReader implements Reader {
       this.r = 0;
       this.w = 0;
       rr = await this.rd.read(this.buf);
-      assert(rr.nread >= 0, "negative read");
-      if (rr.nread === 0) return rr;
-      this.w += rr.nread;
+      const nread = rr === Deno.EOF ? 0 : rr;
```
Comment (piscisaureus) [inline io/bufio.ts]: Ditto

**Comment by piscisaureus** [comment]
Looking great.
@ry since this needs to land in concert with changes to deno (we can't get either to pass CI independently), how should we go about landing this?

I would suggest to make a simple patch that just adds the EOF symbol to deno and nothing else.
Then release that.
Then land this.
Then remove ReadResult from Deno.

**Comment by kt3k** [comment]
@piscisaureus 
> I would suggest to make a simple patch that just adds the EOF symbol to deno and nothing else.
> Then release that.
> Then land this.

This PR depends on the change of Reader interface. Only adding EOF symbol to deno core is not enough for passing the CI of this branch.

I would suggest the following:
1. land https://github.com/denoland/deno/pull/2591
2. release the next version of deno core
3. update deno in this branch (CI should pass here)
4. land this branch

**Comment by ry** [comment]
I've updated to v0.11.0 but the CI is still failing on Windows at `emptyDirIfItExist`
```
AssertionError: Expected function to throw.
    at assertThrowsAsync (file:///D:/a/1/s/testing/asserts.ts:221:15)
    at async emptyDirIfItExist (file:///D:/a/1/s/fs/empty_dir_test.ts:57:9)
    at async runTestsSerial (file:///D:/a/1/s/testing/mod.ts:175:13)
    at async runTests (file:///D:/a/1/s/testing/mod.ts:223:9)
    at async run (file:///D:/a/1/s/test.ts:32:5)
```

**Pull request merged**